program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,375 @@ use crate::c_oracle_header::{
     PC_PRICE_T_COMP_OFFSET,
 };
 use crate::error::OracleError;
+use crate::utils::try_convert;
 use bytemuck::{
     Pod,
     Zeroable,
 };
-use solana_program::msg;
+use std::cmp;
 
+//To make it easy to change the types to allow for more usefull
+//running sums if needed
+use {
+    i64 as SignedTrackerRunningSum,
+    u64 as UnsignedTrackerRunningSum,
+};
+//multiply by this to make up for errors that result when dividing integers
+pub const SMA_TRACKER_PRECISION_MULTIPLIER: i64 = 10;
+
+
+pub trait Tracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
```

Comment:
These shouldn't be generic paramaters, but rather fields in a struct.

Using traits to emulate abstract classes, by implementing common behaviour in their functions, is a code smell that you're trying to use rust in an traditional OOO inheritance-style way. Things will typically be easier if you use a composition approach: a struct which has the common functionality defined on it, with a generic field for the abstract logic.

This trait should be much smaller, containing only the `add_price` method, as that is the only abstract functionality.

Reply:
So it sounds like there are two orthognal comments here.
1. These const generics should not be const generics
2. there's a better way to minimize duplicate code with SMA and TickTrackers

For 1, I disagree without a clearer reason. These things are constant, one of them is litterally the size of an array. So, atleast that IMO belongs as a const genric, so please expand on 1.

For 2, this sounds right, I tried to do that in an earlier draft on my own but things got a little confusing (specially with traits not being able to access members, and the amount of boiler plate code to right getters and setters outweighting the benefits of the split). That said, this implementation has gotten much simpler as I thought more about it, so maybe, it is worth revisiting. Is there any example that you think I should look at it? I am very new to rust, certainly new to this stuff